docker/importer/importer.py

```diff
@@ -313,11 +313,98 @@ def _vuln_ids_from_gcs_blob(self, client: storage.Client,
           strict=self._strict_validation)
     except Exception as e:
       logging.error('Failed to parse vulnerability %s: %s', blob.name, e)
+      # TODO(andrewpollock): I think this needs to be reraised here...
+      # a jsonschema.exceptions.ValidationError only gets raised in strict
+      # validation mode.
       return None
     for vuln in vulns:
       vuln_ids.append(vuln.id)
     return vuln_ids
 
+  def _convert_blob_to_vuln(
+      self, storage_client: storage.Client, ndb_client: ndb.Client,
+      source_repo: osv.SourceRepository, blob: storage.Blob,
+      ignore_last_import_time: bool) -> Optional[Tuple[str]]:
+    """Parse a GCS blob into a tuple of hash and Vulnerability
+
+    Criteria for returning a record:
+    - it's new/modified since last run and the hash has changed
+    - the importer is reimporting the entire source
+      - ignore_last_import_time is True
+    - the record passes OSV JSON Schema validation
+
+    Usually an OSV file has a single vulnerability in it, but it is permissible
+    to have more than one, hence it returns a list of tuples.
+
+    This is runnable in parallel using concurrent.futures.ThreadPoolExecutor
+
+    Args:
+      storage_client: a storage.Client() to use for retrieval of the blob
+      ndb_client: an ndb.Client() to use for Data Store access
+      source_repo: the osv.SourceRepository the blob relates to
+      blob: the storage.Blob object to operate on
+
+    Raises:
+      jsonschema.exceptions.ValidationError when self._strict_validation is True
+      input fails OSV JSON Schema validation
+
+    Returns:
+      a list of one or more tuples of (hash, vulnerability) (from the
+      Vulnerability proto) or None when the blob has an unexpected name
+    """
+    if not _is_vulnerability_file(source_repo, blob.name):
+      return None
+
+    utc_last_update_date = source_repo.last_update_date.replace(
+        tzinfo=datetime.timezone.utc)
+
+    if not ignore_last_import_time and \
+        blob.updated is not None and \
+        not blob.updated > utc_last_update_date:
+      return None
+
+    # The record in GCS appears to be new/changed, examine further.
+    logging.info('Bucket entry triggered for %s/%s', source_repo.bucket,
+                 blob.name)
+
+    # Download in a blob generation agnostic way to cope with the blob
+    # changing between when it was listed and now (if the generation doesn't
+    # match, retrieval fails otherwise).
+    blob_bytes = storage.Blob(
+        blob.name, blob.bucket,
+        generation=None).download_as_bytes(storage_client)
+
+    # This is the atypical execution path (when reimporting is triggered)
+    if ignore_last_import_time:
+      blob_hash = osv.sha256_bytes(blob_bytes)
+      if self._strict_validation:
+        # This may raise a jsonschema.exceptions.ValidationError
+        _ = osv.parse_vulnerabilities_from_data(
```

Comment:
While we're here: we can probably just do this in all cases, because it should be very fast (we're already downloading the Blob), and in the future presumably we want to always do strict validation? 

I'd suggest moving the parsing up to right after the download on line 375, and re-using the return value both here and in the typical case below.

Reply:
Yeah, good call, thanks. I've also DRYed the blob hashing.